### PR TITLE
feat: fetch secret from `IOptionsMonitor` if defined

### LINKED ISSUE #486: [FEAT]: Please support fetching the secret asynchronously
### Describe the need

I need to fetch the secret from Key Vault and it can change over the lifecycle of the container.

The code you have to verify the signature is to fetch the secret at configure time and not asynchronous...

```cs
public static IEndpointConventionBuilder MapGitHubWebhooks(
  this IEndpointRouteBuilder endpoints,
  string path = "/api/github/webhooks",
  string secret = null!
)
```

At startup we are initializing it like this...

```cs
public static WebApplication UseGitHubWebhooks(this WebApplication app)
{
  var githubOptions = app.Services.GetRequiredService<IOptions<GitHubOptions>>().Value;
  var secretClient = app.Services.GetRequiredService<ISecretClient>();
  var secret = secretClient.GetSecret(githubOptions.WebhookSecretKey).ConfigureAwait(false).GetAwaiter().GetResult();
  app.MapGitHubWebhooks($"{ApplicationName.BaseUrl}/v1{githubOptions.WebhookPath}", secret.Value);
  return app;
}
```

This works but there are a couple of problems with this:

1. Having to add an async call at this stage of the application startup isn't great.
2. Theoretically the secret can change in key vault but I'll never be able to update it here without restarting my pods, there is no dynamic fetching of the secret going on.

Now there is a couple of ways that you could solve this but in my opinion simply adding some kind of `IWebhookSecret` interface which has an async method `public Task<string> GetSecret()` would work great.

That way I could inject an instance of an object that knows how to reach out to my `ISecretClient` and dynamically fetch before validating.

```cs
var secret = await secretClient.GetSecret();
if (!await VerifySignatureAsync(context, secret, body).ConfigureAwait(false))
{
  Log.SignatureValidationFailed(logger);
  return;
}
```

### SDK Version

Octokit.net 11.0.1

### API Version

N/A

### Relevant log output

```shell
Not relevant to the API
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by JamieMagee**
Currently, the secret is statically defined when registering the GitHub webhooks endpoint with `MapGitHubWebhooks`. But this means that the `secret` used to validate the webhooks signature is static for the lifetime of the application.

If, instead, we attempt to fetch the secret from `IOptionsMonitor<GitHubWebhookOptions>` it allows us to rotate the secret without restarting the application.

This is also implemented in a backwards compatible manner. If the `secret` parameter passed to `MapGitHubWebhooks` is `null`, but an `IOptionsMonitor<GitHubWebhookOptions>` instance has been registered, the secret will be fetched from there instead.

Closes #486

----

### Before the change?

* The GitHub webhook validation secret is static for the lifetime of the application

### After the change?

* The GitHub webhook 

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [x] No

----

### CODE REVIEW

File: src/Octokit.Webhooks.AspNetCore/GitHubWebhookExtensions.cs [hunk 1]
```diff
@@ -106,7 +120,7 @@ await context.Response.WriteAsync("Payload includes a secret, so the webhook rec
             return false;
         }
 
-        var keyBytes = Encoding.UTF8.GetBytes(secret);
+        var keyBytes = Encoding.UTF8.GetBytes(secret!);
```
Comment (kfcampbell) [inline src/Octokit.Webhooks.AspNetCore/GitHubWebhookExtensions.cs:123]: Should this default to an empty string, or error out if secret is null or empty?

**Review by kfcampbell — APPROVED** [review]

File: src/Octokit.Webhooks.AspNetCore/GitHubWebhookExtensions.cs (hunk 1 above)
Comment (justinmchase) [reply src/Octokit.Webhooks.AspNetCore/GitHubWebhookExtensions.cs:123]: Since the secret is optional, should it just `return true;` if its `null`?

**Review by justinmchase — APPROVED** [review]
Would it be possible to add a little example in the docs of how one could update the secret?

**Comment by JamieMagee** [comment]
@justinmchase I think the official documentation does a better job than I could ever do it. I could link to something like this instead?

https://learn.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-aspnet-core

File: src/Octokit.Webhooks.AspNetCore/GitHubWebhookExtensions.cs (hunk 1 above)
Comment (JamieMagee) [reply src/Octokit.Webhooks.AspNetCore/GitHubWebhookExtensions.cs:123]: It's a little hard to follow, but this is covered with the logic above.

If the secret is `null` or empty then `isSignatureExpected` will be `false`. We cover 3 of the combinations of `isSigned` and `isSignatureExpected` above. By the time we get here we know the event is signed, and we know we are expecting a signature.

File: src/Octokit.Webhooks.AspNetCore/GitHubWebhookOptions.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Octokit.Webhooks.AspNetCore;
+
+public sealed record GitHubWebhookOptions
```
Comment (melotic) [inline src/Octokit.Webhooks.AspNetCore/GitHubWebhookOptions.cs:3]: Should this be sealed? I can imagine a use case (at least personally) where I would like to be able to also lump unrelated configuration to webhooks (like my client secret or app id) and not have to create a separate class for webhooks and GitHub app configurations.

Comment (JamieMagee) [reply src/Octokit.Webhooks.AspNetCore/GitHubWebhookOptions.cs:3]: I sealed it mostly because I seal all of the classes in this library. The performance benefits of the webhook object classes is probably more important than this object. Webhook objects will be constructed on each event, whereas this will be constructed once, on startup.

I can send a follow-up PR to unseal it.
